### PR TITLE
Add support for a post-loaded profile

### PR DESCRIPTION
**Pull request by olysonek**
This is a minimal viable implementation of a post-loaded profile feature. It was implemented to satisfy the needs of the kernel_settings Ansible role.
https://github.com/linux-system-roles/kernel_settings

This feature allows us to apply a profile in a way that is somewhat separate from the way profiles from the `active_profile` file are applied. Namely, the post-loaded profile is always applied after all the profiles from the `active_profile` (or more precisely, the post-loaded profile is always put to the tail of the profile chain when merging). Also, the post-loaded profile setting is not affected by running `tuned-adm profile <some profile>` - the post-loaded profile remains applied.

The name of the post-loaded profile is stored in `/etc/tuned/post_loaded_profile`.

The lack of support for multiple post-loaded profiles was more or less an arbitrary decision. There are no problems with it, as far as I can see. However, there is currently no use case for multiple post-loaded profiles, so I decided to make things easier for our future selves - it's one less thing to worry about if we need to make changes to the profile loading logic. We can add support for multiple post-loaded profiles anytime.

Resolves: rhbz#1798183

Signed-off-by: Ondřej Lysoněk <olysonek@redhat.com>

### CODE REVIEW

**Comment by olysonek** [comment]
TLDR; the following is a list of requirements for the feature and design decisions that I had to make.

The requirements are based on my understanding of the kernel_settings role. @richm Please let me know if they are not accurate.

Also, please let me know if you disagree with any of my decisions or if there's something else that I should consider.

All the requirements and decisions (except for 8.) are tested by the included `runtest.sh`. For easy reference, the test cases are written in order.

1. *Requirement*: After running 'tuned-adm profile <profile name>', the post-loaded profile must remain applied.
2. *Requirement*: Settings from the post-loaded profile must override settings from the profiles set in the active_profile file if there are conflicts.
3. *Requirement*: 'tuned-adm verify' must not fail just because there are conflicts between the active_profile and the post-loaded profile. (The consequence of this is that the post-loaded profile must be merged with the active_profile - meaning, the post-loaded profile is not applied separately.)
4. What should happen on 'tuned-adm off' when a post-loaded profile is applied?
   a) deactivate and clear the post-loaded profile setting on disk
   b) leave the post-loaded profile applied and keep the setting on disk
   c) add a cmdline option to modify the behaviour. Which behaviour should be the default?
5. (Somewhat related to 4.) What should a starting-up Tuned process do if the active_profile setting on disk is empty and the post-loaded profile setting is non-empty?
   a) complain about illegal configuration
   b) apply the post-loaded profile only
   c) not apply any profile
6. 'tuned-adm active' should show the post-loaded profile. But how should it be displayed?
   a) the post-loaded profile should be shown as if it were a regular profile
   b) the post-loaded profile should get shown separately
   c) the post-loaded profile should be shown in both ways at the same time, e.g.:
```
Current active profile: balanced kernel_settings
Post-loaded profile: kernel_settings
```
7. What should 'tuned -p <profile_name>' do?
   a) clear the post-loaded profile setting on disk, set only <profile_name> as the active profile and apply the profile
   b) keep the post-loaded profile setting, set <profile_name> as the active profile and apply both <profile_name> and the post-loaded profile
   c) keep the post-loaded profile setting set on disk, set <profile name> as the active profile and apply only <profile name>
8. (There's no test case for this, but it's implicitly tested by other test cases combined with the fact that the implementation doesn't modify the tuned-adm code.) When Tuned applies a profile, it sends a "profile_changed" signal to DBus. This signal contains the list of applied profiles. Which profiles should be listed in the in the signal when a post-loaded profile is applied?
   a) List all the applied profiles, including the post-loaded one. This would require changes in tuned-adm and possibly other programs that may consume the signal (e.g. Cockpit). ('tuned-adm profile' consumes the signal so that it knows when the profile is fully applied.)
   b) Always list only the active_profile. This will alter the meaning of the signal. If we choose this option, we may want to implement another signal that would contain all the applied profiles, including the post-loaded one.
9. (Related to 1.) On 'tuned-adm profile <profile_name>, Tuned should
   a) reuse the already loaded post-loaded profile name
   b) reload the post-loaded profile name from disk
10. (Related to 1.) Tuned has a habit of writing to the active_profile file after profiles are applied. Should it always write out both the active_profile and post_loaded_profile, or only the one that changed (e.g. due to 'tuned-adm profile')?
    a) write out both. This means that if we change the post_loaded_profile file on disk and then do 'tuned-adm profile', the post_loaded profile will be overwritten.
    b) write out only the one that changed. This might change the semantics of the files slightly - they no longer reflect the currently applied profiles.
11. Should variables be shared between the 'active_profile' profile and the post-loaded profile?
    a) yes
    b) no
   
My choices and the justification:
4. a) seems to be the best choice to preserve the semantics
5. b) so that the kernel_settings profile can be the only applied
   profile.
6. a) seems to be the best choice to preserve the semantics. We can later implement c) if needed.
7. b) is the easiest to implement. No reason other than that.
8. b) because it doesn't break other programs
9. a) because the requirement is that 'tuned-adm profile' does not change the post-loaded profile. Option a) is more in line with that requirement.
10. a) seems to be the best choice to preserve the semantics of the active_profile file. Also, it's the easiest to implement.
11. I think ideally, the variables should not be shared, because (at least for the use case of kernel_settings) the profile hierarchies are unrelated. However, separating the variables could be tricky to implement, so I choose a), unless someone has a strong opinion on this. We could add a config option later that would modify the behaviour.

**Comment by olysonek** [comment]
There were some syntax errors in my previous comment that caused it to display incorrectly. Sorry about that. Hopefully fixed now.

**Comment by olysonek** [comment]
@richm Please review and test.

**Comment by olysonek** [comment]
>  8. b) because it doesn't break other programs

It seems I've implemented something different. I'll fix it.

**Comment by olysonek** [comment]
> > 8. b) because it doesn't break other programs
> 
> It seems I've implemented something different. I'll fix it.

Never mind. I did actually implement 8. b). I just left some confusing dead code lying around.

I added a test case for 8. b) now.

**Comment by richm** [comment]
I'm running tests/beakerlib/bz1798183-RFE-support-post-loaded-profile on Fedora 31.  It passes until it gets to here:
```
:: [ 15:33:18 ] :: [   LOG    ] :: rlServiceStart: Service tuned started successfully
:: [ 15:33:18 ] :: [   PASS   ] :: Command 'rlServiceStart tuned' (Expected 0, got 0)
:: [ 15:33:18 ] :: [  BEGIN   ] :: Running 'tuned-adm off'
:: [ 15:33:18 ] :: [   PASS   ] :: Command 'tuned-adm off' (Expected 0, got 0)
:: [ 15:33:18 ] :: [   PASS   ] :: Check the output of tuned-adm active (Assert: 'No current active profile.' should equal 'No current active profile.')
:: [ 15:33:18 ] :: [   FAIL   ] :: Check that swappiness has not been changed (Assert: '60' should equal '30')
:: [ 15:33:18 ] :: [   FAIL   ] :: Check that dirty ratio has not been changed (Assert: '20' should equal '30')
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
::   Duration: 2s
::   Assertions: 5 good, 2 bad
::   RESULT: FAIL (Check that 'tuned-adm off' causes Tuned to clear the post-loaded profile)
```
not sure if this is a false negative - I'm struggling with the fact that the test restarts tuned several times, and I'm hitting the Fedora 31 systemd service restart limit of 5 times in 10s.  Still investigating, but can you confirm the test works on some version of rhel?

**Comment by olysonek** [comment]
@richm Please try the current version. All of the problems you mentioned should be resolved now. If not, please let me know.

I haven't tested on RHEL yet, but it seems to work on CentOS 8.

**Comment by olysonek** [comment]
@richm You can also find the test results here:
https://artifacts-testing-farm.apps.ci.centos.org/05ed77a6-1171-4213-aae8-3862a6780891/ci/execute.log

There are also results for a couple of other tests in there, so search for `bz1798183-RFE-support-post-loaded-profile`.

**Comment by richm** [comment]
> 10  a) write out both. This means that if we change the post_loaded_profile file on disk and then do tuned-adm profile, the post_loaded_profile file will be overwritten.

So how to update the post_loaded_profile safely?  Is it this?

1) shutdown tuned e.g. systemctl stop tuned
2) edit /etc/tuned/post_loaded_profile
3) start tuned

?

**Comment by olysonek** [comment]
> > 10  a) write out both. This means that if we change the post_loaded_profile file on disk and then do tuned-adm profile, the post_loaded_profile file will be overwritten.
> 
> So how to update the post_loaded_profile safely? Is it this?
>     1. shutdown tuned e.g. systemctl stop tuned
>     2. edit /etc/tuned/post_loaded_profile
>     3. start tuned

Either that, or just edit /etc/tuned/post_loaded_profile and then restart Tuned. That's what I'm doing in the tests.

In the future we could add a new tuned-adm command for changing  the post-loaded profile.

**Comment by olysonek** [comment]
> > > 10  a) write out both. This means that if we change the post_loaded_profile file on disk and then do tuned-adm profile, the post_loaded_profile file will be overwritten.
> > 
> > So how to update the post_loaded_profile safely? Is it this?
> > 1. shutdown tuned e.g. systemctl stop tuned
> > 2. edit /etc/tuned/post_loaded_profile
> > 3. start tuned
> 
> Either that, or just edit /etc/tuned/post_loaded_profile and then restart Tuned. That's what I'm doing in the tests.

You could also send the HUP signal to the Tuned process after writing the file. I'll write a test for this.

**Comment by yarda** [comment]
> 
> My choices and the justification:
> 4. a) seems to be the best choice to preserve the semantics
> 5. b) so that the kernel_settings profile can be the only applied
> profile.
> 6. a) seems to be the best choice to preserve the semantics. We can later implement c) if needed.
> 7. b) is the easiest to implement. No reason other than that.
> 8. b) because it doesn't break other programs
> 9. a) because the requirement is that `tuned-adm profile` does not change the post-loaded profile. Option a) is more in line with that requirement.
> 10. a) seems to be the best choice to preserve the semantics of the `active_profile` file. Also, it's the easiest to implement.
> 11. I think ideally, the variables should not be shared, because (at least for the use case of kernel_settings) the profile hierarchies are unrelated. However, separating the variables could be tricky to implement, so I choose a), unless someone has a strong opinion on this. We could add a config option later that would modify the behaviour.
> 
> EDIT: 8. b) now has a test case.

I mostly agree with your choices, except:
6) I think c) is better, because user should be informed that it's some special profile and they can then start searching what the post-loaded profile is about. Otherwise they could be confused and could spent some hard time trying to get rid of it :)
9) I think the post-loaded profile should be reloaded from the disk as all other profiles are, otherwise users could be surprised of this exceptional behaviour.

**Comment by yarda** [comment]
> > My choices and the justification:
> > 4. a) seems to be the best choice to preserve the semantics
> > 5. b) so that the kernel_settings profile can be the only applied
> > profile.
> > 6. a) seems to be the best choice to preserve the semantics. We can later implement c) if needed.
> > 7. b) is the easiest to implement. No reason other than that.
> > 8. b) because it doesn't break other programs
> > 9. a) because the requirement is that `tuned-adm profile` does not change the post-loaded profile. Option a) is more in line with that requirement.
> > 10. a) seems to be the best choice to preserve the semantics of the `active_profile` file. Also, it's the easiest to implement.
> > 11. I think ideally, the variables should not be shared, because (at least for the use case of kernel_settings) the profile hierarchies are unrelated. However, separating the variables could be tricky to implement, so I choose a), unless someone has a strong opinion on this. We could add a config option later that would modify the behaviour.
> > EDIT: 8. b) now has a test case.
> 
> I mostly agree with your choices, except:
> 6) I think c) is better, because user should be informed that it's some special profile and they can then start searching what the post-loaded profile is about. Otherwise they could be confused and could spent some hard time trying to get rid of it :)
> 9) I think the post-loaded profile should be reloaded from the disk as all other profiles are, otherwise users could be surprised of this exceptional behaviour.

9) a) is not blocker for me (if properly documented), but I think 9) b) would simplify the implementation.

File: tuned/utils/commands.py [hunk 1]
```diff
@@ -471,3 +471,25 @@ def save_active_profile(self, profile_name, manual):
 				f.write(mode + "\n")
 		except (OSError,IOError) as e:
 			raise TunedException("Failed to save profile mode: %s" % e.strerror)
+
```
Comment (yarda) [inline tuned/utils/commands.py:474]: Is there any reason for it to be in the 'commands'? I take it as a hack for storage of utilities/helpers which could be useful in other places of the Tuned code/plugins and these two functions seems to be called just from the daemon.py.

**Comment by olysonek** [comment]
>  9) I think the post-loaded profile should be reloaded from the disk as all other profiles are, otherwise users could be surprised of this exceptional behaviour.

Let me clarify that 9) is about the profile *name* (i.e. the contents of /etc/tuned/post_loaded_profile), not the profile *contents* (i.e. the contents of /etc/tuned/foo/tuned.conf). Did you get that right?

Currently, the name of the post-loaded profile is not reloaded, but the contents are reloaded.

The `switch_profile` method of `Controller` doesn't read any profile names from disk either - it just uses the profile names that were provided.

**Comment by olysonek** [comment]
>  6) I think c) is better, because user should be informed that it's some special profile and they can then start searching what the post-loaded profile is about. Otherwise they could be confused and could spent some hard time trying to get rid of it :)

Yes, I agree. I'll implement c) as soon as I can. By implementing a) first, I was just trying to save some time.

**Comment by yarda** [comment]
> > 1. I think the post-loaded profile should be reloaded from the disk as all other profiles are, otherwise users could be surprised of this exceptional behaviour.
> 
> Let me clarify that 9) is about the profile _name_ (i.e. the contents of /etc/tuned/post_loaded_profile), not the profile _contents_ (i.e. the contents of /etc/tuned/foo/tuned.conf). Did you get that right?
> 
> Currently, the name of the post-loaded profile is not reloaded, but the contents are reloaded.
> 
> The `switch_profile` method of `Controller` doesn't read any profile names from disk either - it just uses the profile names that were provided.

Sorry, I missed that. So it's probably OK then.

File: tuned/utils/commands.py (hunk 1 above)
Comment (olysonek) [reply tuned/utils/commands.py:474]: I wanted it to be in the same place as `save_active_profile`, etc., because those methods are related. In the future, I certainly want to move them some place else. My PR #176 already does that - later I'll update that PR to move the new methods too.

**Comment by olysonek** [comment]
> Sorry, I missed that. So it's probably OK then.

No problem. I guess deciding 9) might be a no-brainer, but I wanted to mention it, because I got bit by it when writing my test.

File: tuned/utils/commands.py (hunk 1 above)
Comment (olysonek) [reply tuned/utils/commands.py:474]: Also, in the future they will be called from admin.py too - I'll need them to implement 6) c) and things like `tuned-adm change-post-loaded-profile`.

**Comment by yarda** [comment]
Thanks, LGTM.

**Comment by yarda** [comment]
Thanks, merging.
